### PR TITLE
[Android] Add NdkToolFinder task & Fix up AOT app build

### PR DESCRIPTION
**Pull request by steveisok**
Since we bumped to NDK 23, having the aot compiler itself generate shared libraries stopped working. This is due to NDK 23 moving most of the toolchain into a common bin directory. AS was left over in each of these directories as an alias to `bin/<triple>-as`.

This change adds a task to collect all of the important NDK toolchain paths. It also fixes up the aot build when `AOTWithLibraryFiles` is set to true and we want the aot compiler to produce shared libraries.

### CODE REVIEW

File: src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                new string[]
+                {
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86), "Android", "android-sdk", "ndk-bundle"),
+                    Path.Combine(Environment.GetFolderPath (Environment.SpecialFolder.ProgramFilesX86), "Android", "android-sdk-windows", "ndk-bundle"),
+                    !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("ProgramW6432"))
+                        ? Path.Combine(Environment.GetEnvironmentVariable("ProgramW6432") ?? "", "Android", "android-sdk", "ndk-bundle")
+                        : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "Android", "android-sdk", "ndk-bundle"),
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Android", "android-sdk", "ndk-bundle"),
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "Android", "android-sdk", "ndk-bundle"),
+                    @"C:\android-sdk-windows"
```
Comment (akoeplinger) [inline src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs:79]: do we really want to support this path?

File: src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                {
+                    ret = path;
+                    break;
+                }
+            }
+
+            return ret;
+        }
+
+        private static List<string?> GetProbingPaths()
```
Comment (akoeplinger) [inline src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs:63]: I'd do the null checks here so you can return `List<string>` and make your life easier in other methods

File: src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            { "linux", "linux-x86_64" },
+            { "windows", "windows-x86_64" }
+        };
+
+        private static readonly Dictionary<string, AndroidArch> validArches = new Dictionary<string, AndroidArch>()
+        {
+            { "arm", new AndroidArch("32-bit ARMv7", "arm", "armeabi-v7a", "arm-linux-androideabi") },
+            { "arm64", new AndroidArch("64-bit ARMv8", "aarch64", "aarch64-v8a", "aarch64-linux-android") },
+            { "x86", new AndroidArch("32-bit Intel", "x86", "x86", "i686-linux-android") },
+            { "x64", new AndroidArch("64-bit Intel", "x86_64", "x86_64", "x86_64-linux-android") }
```
Comment (akoeplinger) [inline src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs:37]: doesn't look like we use the display name (e.g. `64-bit ARMv8`) anywhere so I'd remove it.

File: src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        public string ClangPath
+        {
+            get => clangPath;
+        }
+
+        private void ValidateRequiredProps(string hostOS)
+        {
+            if (Ndk.NdkVersion.Main.Major != 23)
+            {
+                throw new Exception($"NDK 23 is required. An unsupported NDK version was found ({Ndk.NdkVersion.Main.Major}).");
```
Comment (akoeplinger) [inline src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs:95]: is it really the exact version? not >= ?

**Review by akoeplinger — APPROVED** [review]

File: src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs (hunk 4 above)
Comment (steveisok) [reply src/tasks/MobileBuildTasks/Android/Ndk/NdkTools.cs:95]: Yes, for now.  I'll follow up with more broad support.

File: src/mono/sample/Android/AndroidSampleApp.csproj [hunk 5]
```diff
@@ -35,46 +40,54 @@
...
     <MonoAOTCompiler Condition="'$(ForceAOT)' == 'true'"
+        AotModulesTablePath="$(_AotModulesTablePath)"
+        AsPrefix="$(_AsPrefixPath)"
+        Assemblies="@(AotInputAssemblies)"
         CompilerBinaryPath="@(MonoAotCrossCompiler->WithMetadataValue('RuntimeIdentifier','$(TargetOS)-$(TargetArchitecture.ToLowerInvariant())'))"
-        OutputDir="$(_MobileIntermediateOutputPath)"
+        IntermediateOutputPath="$(IntermediateOutputPath)"
+        LibraryFormat="$(_AotLibraryFormat)"
+        LLVMPath="$(MonoAotCrossDir)"
         Mode="Full"
```
Comment (fanyang-mono) [inline src/mono/sample/Android/AndroidSampleApp.csproj:86]: This should be "Normal".

File: src/mono/msbuild/android/build/AndroidBuild.targets [hunk 6]
```diff
@@ -182,17 +215,20 @@
...
         IntermediateOutputPath="$(_MobileIntermediateOutputPath)"
+        LibraryFormat="$(_AotLibraryFormat)"
         LLVMPath="$(MonoAotCrossDir)"
         MibcProfilePath="@(ProfiledAOTProfilePaths)"
         Mode="$(_AOTMode)"
         OutputDir="$(_MobileIntermediateOutputPath)"
-        OutputType="AsmOnly"
+        OutputType="$(_AotOutputType)"
+        ToolPrefix="$(_ToolPrefixPath)"
         UseAotDataFile="$(_UseAotDataFile)"
```
Comment (fanyang-mono) [inline src/mono/msbuild/android/build/AndroidBuild.targets:232]: UseAotDataFile should be false.

**Review by fanyang-mono — COMMENTED** [review]
I was referring to how xamarin-android builded their apps. (https://github.com/xamarin/xamarin-android/blob/03c804262b635089d2d7dab5290eef7d00c66ee2/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.Aot.targets#L109-L130)

File: src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs (hunk 1 above)
Comment (steveisok) [reply src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs:79]: That's what XA checks for as a fallback, so I just lifted it.

File: src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                        ? Path.Combine(Environment.GetEnvironmentVariable("ProgramW6432") ?? "", "Android", "android-sdk", "ndk-bundle")
+                        : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "Android", "android-sdk", "ndk-bundle"),
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Android", "android-sdk", "ndk-bundle"),
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "Android", "android-sdk", "ndk-bundle"),
+                    @"C:\android-sdk-windows"
+                }
+                :
+                new string[]
+                {
+                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Library", "Android", "sdk")
```
[the remark this answers is not in the document]
Comment (steveisok) [reply src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs:84]: Is there a default for linux?  There does not appear to be one in XA either.

**Review by fanyang-mono — APPROVED** [review]

File: src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/tasks/MobileBuildTasks/Android/Ndk/Ndk.cs:84]: There isn't, I was thinking about adding an IsOSX check so we don't try this path which will never exist on Linux

**Pull request merged**